### DISCUSSION TITLE
Meaning of functions called in ipopt

### DISCUSSION THREAD

**Discussion by Srallars97** [casadi/casadi#3178] [General]
Hello everyone,

i am using Casadi in Matlab and trying to optimize it with ipopt. When printing out times, I get the following output, but I can't find anywhere in the internet, what exactly is done in the functions called by ipopt. It is presumably some form of evaluation of the hessian, the function and the gradients.

Does anyone have a closer definition?

Thanks in advance :)

![Bildschirmfoto von 2023-05-10 15-45-19](https://github.com/casadi/casadi/assets/62745750/972b8e7b-dd48-44ef-b0b7-38d5a69cce85)

**Comment by jgillis** [comment]
The canonical for of NLPs is:

```
min          f(x, p)
x

subject to
lbx <=   x    <= ubx
lbg <= g(x, p) <= ubg
```
x: decision variables
p: parameters
f: scalar objective
g: flattened vector of all constraints

x and p are symbols. f, g are expressions dependent on x,p

You may find exact definitions in the source code: https://github.com/casadi/casadi/blob/3.6.2/casadi/interfaces/ipopt/ipopt_interface.cpp#L211

`create_function("nlp_f", {"x", "p"}, {"f"});`
Creates a Function mapping the decision variables x and parameters p to the scalar objective.
`create_function("nlp_g", {"x", "p"}, {"g"});`
Function mapping x and p to the flattened vector of all constraints constraints
`create_function("nlp_grad_f", {"x", "p"}, {"f", "grad:f:x"});`
mapping to objective and gradient of objective wrt x
`create_function("nlp_jac_g", {"x", "p"}, {"g", "jac:g:x"});`
mapping to constraints and constraint Jacobian
`create_function("nlp_hess_l", {"x", "p", "lam:f", "lam:g"},{"triu:hess:gamma:x:x"}, {{"gamma", {"f", "g"}}});`
Computes the Hessian of the Lagrangian $\gamma$ defined with multipliers:
$\gamma = f \lambda_f +  \lambda_g^T g $
$\lambda_f$ is not something you typically find in textbooks. Think of it as an extra scalar scaling parameter for the objective contribution of the Lagrangian.

**Reply by Srallars97** [reply]
Thanks for the response.


I still have the following questions regarding optimization of ipopt.

When running with different prediction-horizons, I get the following plot:

![Bildschirmfoto von 2023-05-22 11-24-37](https://github.com/casadi/casadi/assets/62745750/4385a0f3-8d33-452e-8730-016265faedbc)

On the x-axis is the prediction horizon. On the y-axis is the actual time needed for one solver iteration.

I use ipopt with the following options:
![Bildschirmfoto von 2023-05-22 11-22-35](https://github.com/casadi/casadi/assets/62745750/0c36f78d-8387-4de7-baf4-f7c3eed88d7c)

The time tracked is the time for one entire simulation run with 10000 simulation steps. In these simulation steps ipopt is called 1000 times (every tenth-iteration). 


The following questions:

- What options/options combinations for ipopt can I try to further improve my performance?
- Is there anything else I can do, to further improve the performance of my code?

Thanks in advance :)
